### PR TITLE
extraction latency improvement 

### PR DESCRIPTION
**Pull request by akhilayesantarao**
- added parameter to speed up extraction in update kb request
- added parameter to skip extraction failures in update kb request
- send 10 files at a time to update kb api. All files received in http trigger are put in the queue as one message 
- set operation status for each file accordingly 
- added file extension validation before sending to update kb api

### CODE REVIEW

**Review by somigithub — APPROVED** [review]

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 1]
```diff
@@ -55,65 +62,105 @@ public static class UploadToQnAMaker
...
                         FileUri = fileUri
                     };
-                    msg.Add(queueMessage);
-
+                    msgs.Values.Add(queueMessage);
                     outRecord.Data["status"] = "InQueue";
                     return outRecord;
                 });
-
+            // Add a list of <= 10 files into one message to be extracted as a batch.
```
Comment (jennifermarsman) [inline CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:69]: If the <10 files is a hard perf requirement of what can be handled by the system, let's make sure that we enforce this in the code somewhere.

Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:74]: added a check

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 2]
```diff
@@ -128,20 +183,65 @@ private static async Task<Operation> MonitorOperation(IQnAMakerClient qnaClient,
...
 
-            if (operation.OperationState != OperationStateType.Succeeded)
+        // Checks valid file type before sending for extraction
+        private static bool IsValidFile(string fileName)
+        {
+            HashSet<string> fileExtensions = new HashSet<string> { "tsv", "pdf", "txt", "docx", "xlsx" };
+            string urlExtension;
+            try
             {
-                log.LogError($"Operation {operation.OperationId} failed to completed. ErrorMessage: {operation.ErrorResponse.Error.Message}");
```
Comment (jennifermarsman) [inline CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:134]: Why did you remove this?  I would like the error message for debugging if there are errors.  This also puts it cleanly in the MonitorOperation which would apply it to all operations.

**Review by jennifermarsman — CHANGES_REQUESTED** [review]
Overall, this code looks great.  I have some suggestions.  The biggest thing to fix is a bug in the timing logic so please fix that and update PR before we merge.  Thanks!

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 3]
```diff
@@ -128,20 +183,65 @@ private static async Task<Operation> MonitorOperation(IQnAMakerClient qnaClient,
...
+
+        private static string GetOperationStatus(Operation operation, string fileName, ILogger log)
+        {
+            string operationState = OperationStateType.Succeeded;
+            if (operation.OperationState != OperationStateType.Succeeded && operation.ErrorResponse != null)
+            {
+                var error = operation.ErrorResponse.Error.Details.Where(error => error.Target == fileName);
+                if(error != null && error.Any())
+                {
+                    var errorResponse = error.ToList().First();
```
Comment (jennifermarsman) [inline CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:215]: Add another comment here to explain, as we discussed

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 4]
```diff
@@ -55,65 +62,113 @@ public static class UploadToQnAMaker
...
+                    {
+                        msg.Add(msgBatch);
+                        msgBatch.Values.Clear();
+                    }
                     outRecord.Data["status"] = "InQueue";
                     return outRecord;
                 });
-
+            // Add a list of <= 10 files into one message to be extracted as a batch. 
+            if (msgBatch.Values.Count > 0)
```
[the remark this answers is not in the document]
Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:75]: Couldn't do that. This checks if there are any messages (<10) to be added to the queue after the request records are processed

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs (hunk 2 above)
Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:134]: Done in GetOperationStatus(), added comments

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs (hunk 3 above)
Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs]: done

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 5]
```diff
@@ -55,65 +62,113 @@ public static class UploadToQnAMaker
...
         }
 
         [FunctionName("upload-to-qna-queue-trigger"), Singleton]
         public async static Task Run(
-            [QueueTrigger("upload-to-qna", Connection = "AzureWebJobsStorage")] QnAQueueMessage qnaQueueMessage,
+            [QueueTrigger("upload-to-qna", Connection = "AzureWebJobsStorage")] QnAQueueMessageBatch qnaQueueMessage,
             ILogger log)
         {
-            log.LogInformation("upload-to-qna-queue-trigger: C# Queue trigger function processed");
+            log.LogInformation("upload-to-qna-queue-trigger: C# Queue trigger function processed ");
```
[the remark this answers is not in the document]
Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs]: nope, removed it

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 6]
```diff
@@ -55,65 +62,113 @@ public static class UploadToQnAMaker
                         FileName = blobName,
                         FileUri = fileUri
                     };
-                    msg.Add(queueMessage);
-
+                    msgBatch.Values.Add(queueMessage);
+                    if (msgBatch.Values.Count == 10)
```
[the remark this answers is not in the document]
Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs]: done

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jennifermarsman) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs:75]: You're right.  My mistake, wasn't thinking clearly at 1am.  :)

File: CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs [hunk 7]
```diff
@@ -55,65 +62,113 @@ public static class UploadToQnAMaker
...
+            }
 
-            var updateOp = await qnaClient.Knowledgebase.UpdateAsync(GetAppSetting("KnowledgeBaseID"), updateKB);
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+            // call update KB using REST API
+            var updateOp = await UpdateKB(qnaClient, updateKB, log);
             updateOp = await MonitorOperation(qnaClient, updateOp, log);
+            stopwatch.Stop();
+            log.LogInformation("upload-to-qna-queue-trigger: update operation time = " + stopwatch.Elapsed.Seconds + " seconds. Number of files processed = " + updateKB.Add.Files.Count);
```
[the remark this answers is not in the document]
Comment (akhilayesantarao) [reply CustomSkillForDataIngestion/QnAIntegrationCustomSkill/QnAIntegrationCustomSkill.cs]: Thanks for the explanation, wasn't aware of this. Made the changes

**Review by jennifermarsman — APPROVED** [review]
Thanks for updating.
